### PR TITLE
Update statistics on insert to empty table in Hive connector

### PR DESCRIPTION
**Pull request by wendigo**
Fixes https://github.com/prestosql/presto/issues/2469

This has uncovered side-effect behaviour in Metastore. If we are updating table column statistics in metastore it will synchronize schema from schema.avro.url avro schema to check if columns from stats match actual schema columns. When we unset this attribute in Hive, we are unable to restore previous (dummy) schema.

### CODE REVIEW

File: presto-hive/src/main/java/io/prestosql/plugin/hive/util/Statistics.java [hunk 1]
```diff
@@ -99,12 +99,18 @@ public static HiveBasicStatistics reduce(HiveBasicStatistics first, HiveBasicSta
 
     public static Map<String, HiveColumnStatistics> merge(Map<String, HiveColumnStatistics> first, Map<String, HiveColumnStatistics> second)
     {
-        // only keep columns that have statistics for both sides
-        Set<String> columns = intersection(first.keySet(), second.keySet());
-        return columns.stream()
+        verify(first != null, "first is null");
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/util/Statistics.java]: we usually use `requireNonNull` like here

https://github.com/prestosql/presto/blob/d36b1053d733cfd6d1c2bf043c7ca321eae93da2/presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/BaseJdbcClient.java#L914-L915

File: presto-hive/src/main/java/io/prestosql/plugin/hive/util/Statistics.java [hunk 2]
```diff
@@ -99,12 +99,18 @@ public static HiveBasicStatistics reduce(HiveBasicStatistics first, HiveBasicSta
...
+        verify(second != null, "second is null");
+
+        Set<String> columns = union(first.keySet(), second.keySet());
+        HiveColumnStatistics emptyStats = HiveColumnStatistics.builder().build();
+
+        ImmutableMap<String, HiveColumnStatistics> collect = columns.stream()
                 .collect(toImmutableMap(
                         column -> column,
-                        column -> merge(first.get(column), second.get(column))));
+                        column -> merge(first.getOrDefault(column, emptyStats), second.getOrDefault(column, emptyStats))));
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/util/Statistics.java]: That's not safe in general. It's safe to take empty stats when stats are missing only if you have other basis to conclude the table was empty. eg you could capture row count from basic table stats and do this only if it was 0

File: presto-hive/src/main/java/io/prestosql/plugin/hive/util/Statistics.java [hunk 3]
```diff
@@ -99,12 +99,18 @@ public static HiveBasicStatistics reduce(HiveBasicStatistics first, HiveBasicSta
...
-        // only keep columns that have statistics for both sides
-        Set<String> columns = intersection(first.keySet(), second.keySet());
-        return columns.stream()
+        verify(first != null, "first is null");
+        verify(second != null, "second is null");
+
+        Set<String> columns = union(first.keySet(), second.keySet());
+        HiveColumnStatistics emptyStats = HiveColumnStatistics.builder().build();
+
+        ImmutableMap<String, HiveColumnStatistics> collect = columns.stream()
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/util/Statistics.java]: ImmutableMap -> Map (we typically declare Maps and Lists with the interface)

File: presto-hive/src/main/java/io/prestosql/plugin/hive/util/Statistics.java [hunk 4]
```diff
@@ -117,63 +123,74 @@ public static HiveColumnStatistics merge(HiveColumnStatistics first, HiveColumnS
...
                     reduce(first.get().getFalseCount(), second.get().getFalseCount(), ADD, false)));
         }
-        return Optional.empty();
+
+        return firstPresentOrEmpty(first, second);
+    }
+
+    private static <T> Optional<T> firstPresentOrEmpty(Optional<T> left, Optional<T> right)
+    {
+        verify(!(left.isPresent() && right.isPresent()), "either left optional or right should not be present");
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/util/Statistics.java]: Why?

method name suggests it should return the first present, even if both are present

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 5]
```diff
@@ -4053,6 +4053,32 @@ public void testCollectColumnStatisticsOnInsert()
...
+    @Test
+    public void testCollectColumnStatisticsOnInsertToEmptyTable()
+    {
+        String tableName = "test_collect_column_statistics_empty_table";
+
+        assertUpdate(format("" +
+                "CREATE TABLE %s ( " +
+                "   col INT " +
+                ") ",
+                tableName));
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]: you can put all this on one line

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 6]
```diff
@@ -4053,6 +4053,32 @@ public void testCollectColumnStatisticsOnInsert()
         assertUpdate("DROP TABLE " + tableName);
     }
 
+    @Test
+    public void testCollectColumnStatisticsOnInsertToEmptyTable()
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]: add a test where the table is empty and it's `ANALYZE`-d, but on one column (of two) only

File: presto-hive/src/main/java/io/prestosql/plugin/hive/util/Statistics.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (martint) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/util/Statistics.java]: Actually, in this case, you can just return the result without the intermediate variable

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java (hunk 6 above)
Comment (wendigo) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]: Done.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/BridgingHiveMetastore.java [hunk 7]
```diff
@@ -116,9 +116,9 @@ public PartitionStatistics getTableStatistics(HiveIdentity identity, Table table
     }
 
     @Override
-    public void updateTableStatistics(HiveIdentity identity, String databaseName, String tableName, Function<PartitionStatistics, PartitionStatistics> update)
+    public void updateTableStatistics(HiveIdentity identity, Table table, Function<PartitionStatistics, PartitionStatistics> update)
     {
-        delegate.updateTableStatistics(identity, databaseName, tableName, update);
+        delegate.updateTableStatistics(identity, toMetastoreApiTable(table), update);
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/BridgingHiveMetastore.java]: As discussed offline. I am not sure where it's ok to repopulate metastore based on `toMetastoreApiTable` results.

@electrum are we losing information?

Comment (findepi) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/BridgingHiveMetastore.java]: @wendigo , i talked with @electrum about this. Conclusion was -- this is not safe to persist a roundrip of `fromMetastoreApiTable + toMetastoreApiTable`. 
Instead, we should find a workaround for Avro tables that would get us going. Or do additional call to metastore for Avro tables.

Let's discuss further.
